### ISSUE TITLE
Add `Dict` and `Set` primitive for Map, HashMap, IntMan, Set, HashSet and IntSet

### ISSUE THREAD

**Issue by ChShersh** [Holmusk/elm-street#26]
Elm has standard data types called `Dict` and `Set` for container packages:

* https://package.elm-lang.org/packages/elm-lang/core/5.1.1/Dict
* https://package.elm-lang.org/packages/elm-lang/core/5.1.1/Set

It's probably a good idea to support them too for common Haskell containers.

**Comment by arbus** [comment]
Some updates on this issue:

I have taken a stab at implementing this and my wip can be found in the branch `arbus/26-Add-Dict-and-Set-primitive`

This is straightforward for types of `(Elm a) => Map Text a` but a serious pain for anything else. Even for straightforward enum types that would normally serialize to just plain text, it can't be decoded on the elm side because elm requires its `Dict` types' keys to be of type `comparable` which is a special `Ord,Eq` constraint which only built-in types adhere to and we cannot extend it to support our custom types as well.

I'm not sure how useful the idea of implementing this only in the scenerio where the map key is a `Text` is useful.

**Comment by chshersh** [comment]
@arbus Am I understanding correctly that the main problem here is that you can't have data types like `Map MyEnum a` or `Set MyEnum` in Elm because enums in Elm doesn't have required built-in interfaces implemented? So even if you can write decoder and encoder for such `Map`, you can't use it in Elm code, it will be a compile-time error, right?

**Comment by arbus** [comment]
Yes that is correct, only things that are `comparable` can be the key of a dict or be put in a set

**Comment by chshersh** [comment]
@arbus Okay, in that case I propose to put this issue on hold and keep your branch. It contains useful code for this feature. I also think that using `Map Text a` on frontend is awkward and will complicate our code. Maybe somebody will implement `enum-map` package with JSON encoders and decoders. Or maybe `elm-0.20` will provide these builtin instances for enums. Who know in what ways our lives will be improved...

**Comment by turboMaCk** [comment]
For the record these 2 packages (of mine, sorry for shameless promotion) are solving problem with `comparable`

- [any-dict](https://package.elm-lang.org/packages/turboMaCk/any-dict/latest/)
- [any-set](https://package.elm-lang.org/packages/turboMaCk/any-set/latest/)

Anyway, generating code for this would be still challenging become it would require generating `key -> comparable` functions.

My suggestion would be to:

- Generate Set for `IntSet`
- Generate Dist for `IntMap`
- Generate `List` for `Set`
- Generate `List (key, value)` for Map, Hashmap and AssocList.

**Comment by silky** [comment]
Elm really made life unpleasant with this decision about comparable :tear:.
